src/main/java/com/fasterxml/jackson/databind/deser/std/JsonNodeDeserializer.java

```diff
@@ -755,13 +755,15 @@ protected final JsonNode _fromFloat(JsonParser p, DeserializationContext ctxt,
             return _fromBigDecimal(ctxt, nodeFactory, p.getDecimalValue());
         }
         if (ctxt.isEnabled(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS)) {
-            // [databind#4194] Add an option to fail coercing NaN to BigDecimal
-            // Currently, Jackson 2.x allows such coercion, but Jackson 3.x will not
-            if (p.isNaN() && ctxt.isEnabled(JsonNodeFeature.FAIL_ON_NAN_TO_BIG_DECIMAL_COERCION)) {
-                ctxt.handleWeirdNumberValue(handledType(), p.getDoubleValue(),
-                        "Cannot convert NaN into BigDecimal");
-            }
             try {
+                // [databind#4194] Add an option to fail coercing NaN to BigDecimal
+                // Currently, Jackson 2.x allows such coercion, but Jackson 3.x will not.
+                // And this conditional check is inside try-catch block because `isNan()` can be
+                // unreliable, as of 2.17 [jackson-core#1135]
+                if (p.isNaN() && ctxt.isEnabled(JsonNodeFeature.FAIL_ON_NAN_TO_BIG_DECIMAL_COERCION)) {
+                    ctxt.handleWeirdNumberValue(handledType(), p.getDoubleValue(),
+                            "Cannot convert NaN into BigDecimal");
+                }
```

Comment:
this is not what I meant - p.isNan is unreliable and should not be used

this code code can be changed to remove this and put this in the catch

```
if (ctxt.isEnabled(JsonNodeFeature.FAIL_ON_NAN_TO_BIG_DECIMAL_COERCION)) {
                    ctxt.handleWeirdNumberValue(handledType(), p.getDoubleValue(),
                            "Cannot convert NaN into BigDecimal");
          return; //??
} else {
   // fall through
}
``` 


If you give me a week or 2, I can get back to this and  try to fix it myself.

Reply:
First thing would be to show how `JsonNode` serialization would be compromised by this change.
Is there a case where `BigDecimal` value would now be incorrect?

Reply:
@JooHyukKim As usual, it'd be necessary to have a failing case first before trying to solve a problem. Otherwise we have no verification of the fix, or check against regression.

And specifically, as far as I can see, `p.isNaN()` never throws an exception; at least for JSON parser. So enclosing that in try-catch block would serve no purpose.

Reply:
Hmmh. Ok, I think I know what is likely problematic... Double value overflow becoming `+INFINITY`, in case of `JsonNodeFeature.FAIL_ON_NAN_TO_BIG_DECIMAL_COERCION` being true.
So this would be actual bug for this handling.

But we do need `jackson-core` test reproduction. Maybe:

https://github.com/FasterXML/jackson-core/pull/1135

does that, I'll have another look.

Reply:
Ok. So one issue relevant here is this:

1. JsonParser.isNaN() immediately after `parser.nextToken()` is accurate; only considering non-standard values (if allowed)
2. JsonParser.getNumberType() coerces all fp values into NumberType.DOUBLE and thus Double value overflow can induce NaNs
3. JsonNodeDeserializer should not call `getNumberType()` for JSON -- but! It should call it for binary formats that DO provide this information

so it's combo of (2) and (3) that are problematic.

Reply:
I agree we need a reproduction. I will close this, because I think I missed lots of information here thus need to catch up on what's missing. 

My mistake 🙏🏼 Thank you for explaining things @pjfanning @cowtowncoder !